File: projectBlueWater/src/main/java/com/lasthopesoftware/resources/loopers/specs/GivenMyThreadName/WhenStartingALooperOnTheThread.java
package com.lasthopesoftware.resources.loopers.specs.GivenMyThreadName;

import android.os.Handler;
import android.os.Looper;
import com.lasthopesoftware.bluewater.shared.promises.extensions.specs.FuturePromise;
import com.lasthopesoftware.resources.loopers.HandlerThreadCreator;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.util.concurrent.CountDownLatch;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.robolectric.Shadows.shadowOf;

@RunWith(RobolectricTestRunner.class)
public class WhenStartingALooperOnTheThread {

	private Looper looper;

	@Before
	public void context() throws Throwable {
		looper = new FuturePromise<>(HandlerThreadCreator.promiseNewHandlerThread("MyThreadName", 3)).get().getLooper();

		final Handler handler = new Handler(looper);
		final CountDownLatch countDownLatch1 = new CountDownLatch(1);
		handler.post(countDownLatch1::countDown);

		shadowOf(looper).getScheduler().advanceToLastPostedRunnable();
		countDownLatch1.await();
	}

	@Test
	public void thenTheLooperThreadNameIsCorrect() {
		assertThat(looper.getThread().getName()).isEqualTo("MyThreadName");
	}

	@Test
	public void thenTheLooperThreadIsStarted() {
		assertThat(looper.getThread().isAlive()).isTrue();
	}
}

File: integrationTests/src/test/kotlin/com/lasthopesoftware/resources/loopers/GivenMyThreadName/WhenStartingALooperOnTheThread.kt
package com.lasthopesoftware.resources.loopers.GivenMyThreadName

import android.os.Handler
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.lasthopesoftware.resources.loopers.HandlerThreadCreator
import org.assertj.core.api.AssertionsForClassTypes
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Shadows.shadowOf
import java.util.concurrent.CountDownLatch

@RunWith(AndroidJUnit4::class)
class WhenStartingALooperOnTheThread {

	companion object {
		val looper = lazy {
			val looper = HandlerThreadCreator.promiseNewHandlerThread("MyThreadName", 3).toExpiringFuture().get()!!.looper
			val handler = Handler(looper)
			val countDownLatch = CountDownLatch(1)
			handler.post { countDownLatch.countDown() }
			val shadowLooper = shadowOf(looper)
			shadowLooper.isPaused = true
			shadowLooper.idle()
			countDownLatch.await()
			looper
		}
	}

	@Test
	fun thenTheLooperThreadNameIsCorrect() {
		AssertionsForClassTypes.assertThat(looper.value.thread.name).isEqualTo("MyThreadName")
	}

	@Test
	fun thenTheLooperThreadIsStarted() {
		AssertionsForClassTypes.assertThat(looper.value.thread.isAlive).isTrue
	}
}
